File: Anagram.java
//Time Complexity : O(nklogk)-n no of string , klogk sort strings 
//Space Complexity :O(n)

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
       if (strs == null || strs.length == 0) {
            return new ArrayList<>();
        } 
         Map<String, List<String>> map = new HashMap<>();
          for (String str : strs) {
            char[] charArray = str.toCharArray();
            Arrays.sort(charArray); //klogk
            String sortedStr = new String(charArray);
            
            if (!map.containsKey(sortedStr)) {
                map.put(sortedStr, new ArrayList<>());
            }
            map.get(sortedStr).add(str);
        }
        
        return new ArrayList<>(map.values());  
    }
}




///Approach 2
// Time Complexity : O(nK), n= length of strs, K = length of longest string

// Space Complexity : O(n)
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
       if (strs == null || strs.length == 0) {
            return new ArrayList<>();
        } 
         Map<Double, List<String>> map = new HashMap<>();
          for (String s : strs) {
            double pp = PrimeProduct(s);
            if(!map.containsKey(pp)){
                map.put(pp,new ArrayList<String>());
            }
            map.get(pp).add(s);
        }
        
        return new ArrayList<>(map.values());  
    }


private double PrimeProduct(String s){
        int[] primeNum = {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97,101,103};
        double result = 1;
        for(int i=0;i<s.length();i++){
            char ch = s.charAt(i);

            result = result * primeNum[ch-'a'];
        }

        return result;
    }
}

File: Anagram.py
# Time Complexity: O(m*nlogn) --> for sorting nlogn, inserting to hashmap m
# Space Complexity: O(nm)
# Did this code successfully run on Leetcode: Yes
# Any problem you faced while coding this: No

def group_anagrams(strs):
    map = {}
    for s in strs:
        sorted_str = ''.join(sorted(s))
        if sorted_str not in map:
            map[sorted_str] = [s]
        else:
            map[sorted_str].append(s)
    return map.values()


# Example 1
strs1 = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(group_anagrams(strs1))

# Example 2
strs2 = [""]
print(group_anagrams(strs2))

# Example 3
strs3 = ["a"]
print(group_anagrams(strs3))
